File: scripts/issue_license.py
#!/usr/bin/env python3
"""
PromptBar license issuer, Python port for Linux (Raspberry Pi etc).

Reads the same private key file the Swift issuer writes
(scripts/license-private.key, raw 32-byte Ed25519 in base64) and produces
identical JSON to scripts/issue-license.swift. Cross-verifies against the
public key embedded in Helpers/LicenseValidator.swift.

First time setup:
    pip3 install cryptography
    # Generate the keypair via the Swift script on your Mac, then copy
    # scripts/license-private.key to the Pi at the same path.

Usage:
    python3 scripts/issue_license.py <email> <order-id> [min-version]

Outputs the signed license JSON on stdout. Redirect to a file:
    python3 scripts/issue_license.py buyer@example.com ko-fi-abc > buyer.promptbar
"""

import json
import os
import sys
from datetime import datetime, timezone
from base64 import b64decode, b64encode

try:
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
except ImportError:
    print("ERROR: cryptography package missing. Install with: pip3 install cryptography",
          file=sys.stderr)
    sys.exit(1)


def canonical_json(obj: dict) -> bytes:
    """Match Swift's JSONSerialization with .sortedKeys exactly.

    Swift's sortedKeys output:
      - keys sorted lexicographically
      - no whitespace, separators are ',' and ':'
    Python: json.dumps(obj, separators=(',', ':'), sort_keys=True)
    """
    return json.dumps(obj, separators=(',', ':'), sort_keys=True).encode('utf-8')


def load_private_key(path: str) -> Ed25519PrivateKey:
    with open(path, 'r') as f:
        raw_b64 = f.read().strip()
    raw = b64decode(raw_b64)
    if len(raw) != 32:
        raise ValueError(f"Expected 32-byte Ed25519 key, got {len(raw)} bytes from {path}")
    return Ed25519PrivateKey.from_private_bytes(raw)


def issue(email: str, order_id: str, min_version: str = "2.0.0",
          private_key_path: str = "scripts/license-private.key") -> dict:
    key = load_private_key(private_key_path)

    issued_at = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

    canonical = {
        "email": email.lower().strip(),
        "issued_at": issued_at,
        "order_id": order_id.strip(),
        "product": "PromptBar",
        "min_version": min_version,
    }

    payload = canonical_json(canonical)
    signature = key.sign(payload)

    license_blob = dict(canonical)
    license_blob["signature"] = b64encode(signature).decode('ascii')
    return license_blob


def main():
    if len(sys.argv) < 3:
        print(__doc__, file=sys.stderr)
        sys.exit(1)
    email = sys.argv[1]
    order_id = sys.argv[2]
    min_version = sys.argv[3] if len(sys.argv) >= 4 else "2.0.0"

    private_key_path = os.environ.get(
        "PROMPTBAR_PRIVATE_KEY",
        "scripts/license-private.key"
    )

    license_blob = issue(email, order_id, min_version, private_key_path)
    # Pretty-printed JSON matches the Swift script's output style.
    print(json.dumps(license_blob, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: scripts/issue-license.swift
#!/usr/bin/env xcrun --sdk macosx swift

import Foundation
import CryptoKit

// PromptBar license issuer.
//
// First time: generate the keypair.
//
//   swift scripts/issue-license.swift --keygen
//
// The script writes scripts/license-private.key (gitignored) and prints the
// public key. Paste the public key string into PUBLIC_KEY_BASE64 inside
// Helpers/LicenseValidator.swift.
//
// Per customer: issue a signed license.
//
//   swift scripts/issue-license.swift <email> <kofi-order-id> [min-version] > license.promptbar
//
// Email the resulting .promptbar file to the buyer.

let args = CommandLine.arguments
let privateKeyPath = "scripts/license-private.key"

if args.count > 1 && args[1] == "--keygen" {
    let key = Curve25519.Signing.PrivateKey()
    let privateB64 = key.rawRepresentation.base64EncodedString()
    let publicB64 = key.publicKey.rawRepresentation.base64EncodedString()
    try privateB64.write(toFile: privateKeyPath, atomically: true, encoding: .utf8)
    FileManager.default.createFile(atPath: privateKeyPath, contents: Data(privateB64.utf8))
    try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: privateKeyPath)
    print("Private key written to \(privateKeyPath) (mode 600).")
    print("DO NOT commit it. It is gitignored.")
    print("")
    print("Public key (paste into PUBLIC_KEY_BASE64 in LicenseValidator.swift):")
    print(publicB64)
    exit(0)
}

guard
    let privateData = try? String(contentsOfFile: privateKeyPath, encoding: .utf8),
    let rawKey = Data(base64Encoded: privateData.trimmingCharacters(in: .whitespacesAndNewlines)),
    let privateKey = try? Curve25519.Signing.PrivateKey(rawRepresentation: rawKey)
else {
    FileHandle.standardError.write(Data("ERROR: \(privateKeyPath) missing or invalid. Run: swift scripts/issue-license.swift --keygen\n".utf8))
    exit(1)
}

guard args.count >= 3 else {
    FileHandle.standardError.write(Data("Usage: swift scripts/issue-license.swift <email> <order-id> [min-version]\n".utf8))
    exit(1)
}

let email = args[1].lowercased()
let orderID = args[2]
let minVersion = args.count >= 4 ? args[3] : "2.0.0"

let iso = ISO8601DateFormatter()
iso.formatOptions = [.withInternetDateTime]
let issuedAt = iso.string(from: Date())

// Canonical payload (signature is computed over this exact JSON byte sequence,
// keys sorted, no whitespace).
let canonical: [String: Any] = [
    "email": email,
    "issued_at": issuedAt,
    "order_id": orderID,
    "product": "PromptBar",
    "min_version": minVersion
]

guard let payload = try? JSONSerialization.data(withJSONObject: canonical, options: [.sortedKeys]) else {
    FileHandle.standardError.write(Data("ERROR: failed to serialize payload\n".utf8))
    exit(1)
}

let signature: Data
do {
    signature = try privateKey.signature(for: payload)
} catch {
    FileHandle.standardError.write(Data("ERROR: signing failed: \(error)\n".utf8))
    exit(1)
}

let license: [String: Any] = [
    "email": email,
    "issued_at": issuedAt,
    "order_id": orderID,
    "product": "PromptBar",
    "min_version": minVersion,
    "signature": signature.base64EncodedString()
]

if let out = try? JSONSerialization.data(withJSONObject: license, options: [.prettyPrinted, .sortedKeys]),
   let text = String(data: out, encoding: .utf8) {
    print(text)
}
